pkg/clusteragent/autoscaling/controller.go

```diff
@@ -0,0 +1,148 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+package autoscaling
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// Processor defines the interface that needs to be implemented by the controller
+type Processor interface {
+	Process(key, ns, name string)
+}
+
+// Controller is generic controller that can be used to implement a controller for a specific resource
+type Controller struct {
+	processor Processor
+	synced    cache.InformerSynced
+	context   context.Context
+
+	// Fields available to child controllers
+	ID        string
+	Client    dynamic.Interface
+	Lister    cache.GenericLister
+	Workqueue workqueue.RateLimitingInterface
+	IsLeader  func() bool
+}
+
+// NewController returns a new workload autoscaling controller
+func NewController(
+	controllerID string,
+	processor Processor,
+	client dynamic.Interface,
+	informer dynamicinformer.DynamicSharedInformerFactory,
+	gvr schema.GroupVersionResource,
+	isLeader func() bool,
+	observable Observable,
+) (*Controller, error) {
+	mainInformer := informer.ForResource(gvr)
+	c := &Controller{
+		processor: processor,
+		ID:        controllerID,
+		Client:    client,
+		Lister:    mainInformer.Lister(),
+		synced:    mainInformer.Informer().HasSynced,
+		Workqueue: workqueue.NewRateLimitingQueue(workqueue.DefaultItemBasedRateLimiter()),
```

Comment:
I think we can configure the workqueue to have some telemetry. Could we use this ?

Reply:
It's indeed definitely possible, but we can implement that later.

Reply:
@adel121 recently implemented a [QueueMetricsProvider](https://github.com/DataDog/datadog-agent/blob/3316e3710d8dc78c97ed4fcb2ada6739a590c220/pkg/util/workqueue/telemetry/telemetry.go#L60). I think you can use that and it would be a very small change